examples/data-objects/inventory-app/src/legacySharedTreeInventoryList.ts

```diff
@@ -0,0 +1,136 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	Change,
+	ChangeNode,
+	Definition,
+	SharedTree as LegacySharedTree,
+	SharedTreeEvent,
+	StablePlace,
+	TraitLabel,
+} from "@fluid-experimental/tree";
+import { DataObject, DataObjectFactory } from "@fluidframework/aqueduct";
+import { IFluidHandle } from "@fluidframework/core-interfaces";
+
+import { IInventoryListUntyped, IPart } from "./interfaces";
+
+const legacySharedTreeKey = "legacySharedTree";
+
+/**
+ * An inventory list implemented using LegacySharedTree.
+ */
+export class LegacySharedTreeInventoryList extends DataObject implements IInventoryListUntyped {
+	private _tree: LegacySharedTree | undefined;
+	private get tree() {
+		if (this._tree === undefined) {
+			throw new Error("Not initialized properly");
+		}
+		return this._tree;
+	}
+
+	protected async initializingFirstTime() {
+		const legacySharedTree = this.runtime.createChannel(
+			undefined,
+			LegacySharedTree.getFactory().type,
+		) as LegacySharedTree;
+
+		// Initialize the inventory with two parts at zero quantity
+		// REV: Is ChangeNode appropriate here? Or should I use some other type (BuildNode?)?
+		const inventoryNode: ChangeNode = {
+			identifier: legacySharedTree.generateNodeId(),
+			// REV: This is based on the Definition casting from Bubblebench, it doesn't really seem
+			// to matter what is used here though.  Is there a better way to omit the definition here?
+			definition: "array" as Definition,
+			traits: {
+				nut: [
+					{
+						identifier: legacySharedTree.generateNodeId(),
+						definition: "scalar" as Definition,
+						traits: {},
+						payload: 0,
+					},
+				],
+				bolt: [
+					{
+						identifier: legacySharedTree.generateNodeId(),
+						definition: "scalar" as Definition,
+						traits: {},
+						payload: 0,
+					},
+				],
+			},
+		};
+		legacySharedTree.applyEdit(
+			Change.insertTree(
+				inventoryNode,
+				StablePlace.atStartOf({
+					parent: legacySharedTree.currentView.root,
+					label: "parts" as TraitLabel,
+				}),
+			),
+		);
+
+		this.root.set(legacySharedTreeKey, legacySharedTree.handle);
+	}
+
+	protected async hasInitialized() {
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		this._tree = await this.root
+			.get<IFluidHandle<LegacySharedTree>>(legacySharedTreeKey)!
+			.get();
+
+		// REV: I see the tsdoc for EditCommitted recommends using a Checkout instead but that seems
+		// like overkill (plus I don't see how it relates exactly).  Is there a better way to just observe
+		// that the scalars are changing and raise "inventoryChanged"?
+		this._tree.on(SharedTreeEvent.EditCommitted, () => {
+			this.emit("inventoryChanged");
+		});
+	}
+
+	// REV: Lots of casts and non-null assertions below - is there a more type-confident way to do this?
```

Comment:
maybe a little, but not too much: the lack of a schema system is one of the more painful things to deal with in legacy tree. The `TreeNodeHandle` API helps some with ergonomics of the API, but doesn't really fix the typing.

Simpler domain models might do well to write some boilerplate to hide the lower-level APIs (which is basically putting a poor-man's schema system on). That's what I did for a [pet project](https://github.com/Abe27342/belle-puzzles/blob/main/packages/puzzlehunt-model/src/puzzlehunt.ts) written on legacy shared-tree for example, and it works OK but doesn't scale very well.